packages/data-api/src/fetchData.js

```diff
@@ -6,63 +6,159 @@ import { utcMoment } from '@tupaia/utils';
+    case 'FINAL_EACH_YEAR':
+      return `GROUP BY ${COMMON_FIELDS}, year_period`;
+    case 'MOST_RECENT':
+      return `GROUP BY ${COMMON_FIELDS}`;
+    default:
+      return '';
+  }
+};
+
+const getA2FinalWhereClause = firstAggregationType => {
+  switch (firstAggregationType) {
+    case 'FINAL_EACH_DAY':
+      return 'AND day_period = a1.day_period';
+    case 'FINAL_EACH_WEEK':
+      return 'AND week_period = a1.week_period';
+    case 'FINAL_EACH_MONTH':
+      return 'AND month_period = a1.month_period';
+    case 'FINAL_EACH_YEAR':
+      return 'AND year_period = a1.year_period';
+    default:
+      return '';
+  }
+};
+
+const getA1Select = firstAggregationType => {
+  switch (firstAggregationType) {
+    case 'FINAL_EACH_DAY':
+      return `SELECT ${COMMON_FIELDS}, day_period`;
+    case 'FINAL_EACH_WEEK':
+      return `SELECT ${COMMON_FIELDS}, week_period`;
+    case 'FINAL_EACH_MONTH':
+      return `SELECT ${COMMON_FIELDS}, month_period`;
+    case 'FINAL_EACH_YEAR':
+      return `SELECT ${COMMON_FIELDS}, year_period`;
+    case 'MOST_RECENT':
+      return `SELECT ${COMMON_FIELDS}`;
+    default:
+      return `SELECT ${COMMON_FIELDS}, ${ANSWER_SPECIFIC_FIELDS}`;
+  }
+};
+
+const getA2Join = firstAggregationType => {
+  return SUPPORTED_AGGREGATIONS.includes(firstAggregationType)
+    ? `CROSS JOIN LATERAL (
+      SELECT ${ANSWER_SPECIFIC_FIELDS}
+      FROM analytics
+      WHERE entity_code = a1.entity_code
+      AND data_element_code = a1.data_element_code
+      ${getA2FinalWhereClause(firstAggregationType)}
+      order by date desc
+      limit 1
+    ) as a2`
+    : '';
+};
+
+const generateBaseSqlQuery = ({
+  dataElementCodes,
+  organisationUnitCodes,
+  surveyCode,
+  eventId,
+  startDate,
+  endDate,
+  aggregations,
+}) => {
+  const firstAggregationType = aggregations && aggregations[0] && aggregations[0].type;
   const sqlQuery = new SqlQuery(
     `
-    SELECT
-      survey_response.id as "surveyResponseId",
-      survey_response.submission_time as "date",
-      entity.code as "entityCode",
-      entity.name as "entityName",
-      question.code as "dataElementCode",
-      question.type as "type",
-      answer.text as "value"
-    FROM
-      survey_response
-    JOIN
-      answer ON answer.survey_response_id = survey_response.id
-    JOIN
-      entity ON entity.id = survey_response.entity_id
-    JOIN
-      question ON question.id = answer.question_id
-    JOIN
-      survey ON survey.id = survey_response.survey_id
-    WHERE
-      question.code IN ${SqlQuery.parameteriseArray(dataElementCodes)}
-    AND
-      entity.code IN ${SqlQuery.parameteriseArray(organisationUnitCodes)}
+    SELECT 
+      date AS "date",
+      entity_code AS "entityCode",
+      entity_name AS "entityName",
+      data_element_code AS "dataElementCode",
+      event_id AS "eventId",
+      value AS "value",
+      answer_type AS "type"
+    FROM (
+      ${getA1Select(firstAggregationType)}
```

Comment:
I think `getA1Select` returns different fields based on the `firstAggregationType`, for some aggregations it only returns the COMMON_FIELDS which are `data_element_code, entity_code, entity_name`. So it might not return all the columns asked in the outer select part above (listed below). Would that throw an error or something?

```
SELECT 
date AS "date",
entity_code AS "entityCode",
entity_name AS "entityName",
data_element_code AS "dataElementCode",
event_id AS "eventId",
value AS "value",
answer_type AS "type"
```

Reply:
So, the rest of those fields are returned by the `a2 select`, which should be joined if the relevant aggregation is being used. If there are fields missing though, you are right an SQL error would be thrown I believe.